File: src/compare_rules.py
"""Shared compare query helpers."""

from __future__ import annotations

import re
from typing import Callable

COMPARE_QUERY_RE = re.compile(r"(比較|差異|不同|對比|比對|\bvs\b|\bversus\b)", re.IGNORECASE)


def normalize_text(value: object) -> str:
    return str(value or "").strip()


def is_compare_like_query(text: object) -> bool:
    normalized = normalize_text(text)
    if not normalized:
        return False
    return bool(COMPARE_QUERY_RE.search(normalized))


def should_prefer_wifi_compare(text: object, is_wifi_specific_query: Callable[[object], bool] | None) -> bool:
    if not callable(is_wifi_specific_query):
        return False
    return is_compare_like_query(text) and bool(is_wifi_specific_query(text))

File: frontend/lib/compare-rules.js
const COMPARE_QUERY_RE = /(比較|差異|不同|對比|比對|\bvs\b|\bversus\b)/i

export function normalizeText(value) {
  return String(value || '').trim()
}

export function isCompareLikeQuery(text) {
  const normalized = normalizeText(text)
  if (!normalized) return false
  return COMPARE_QUERY_RE.test(normalized)
}

export function shouldPreferWifiCompare(text, isWifiSpecificQuery) {
  if (typeof isWifiSpecificQuery !== 'function') return false
  return isCompareLikeQuery(text) && isWifiSpecificQuery(text)
}

const compareRules = {
  normalizeText,
  isCompareLikeQuery,
  shouldPreferWifiCompare,
}

if (typeof window !== 'undefined') {
  window.KBCompareRules = compareRules
}

export default compareRules
